simulation/src/simulation.rs

```diff
@@ -0,0 +1,362 @@
+            recorder,
+            endpoints: self.endpoints,
+            delay_between_requests: Duration::from_secs_f64(1. / requests_per_second),
+        }
+    }
+}
+
+pub struct SimulationBuilder2 {
+    runtime: Runtime,
+    metrics: Arc<MetricRegistry>,
+    builder: Builder<SimulationRawClientBuilder>,
+    recorder: Arc<Mutex<SimulationMetricsRecorder>>,
+    endpoints: Vec<Endpoint>,
+    delay_between_requests: Duration,
+}
+
+impl SimulationBuilder2 {
+    pub fn send_until(self, cutoff: Duration) -> SimulationBuilder3 {
+        let num_requests = cutoff.as_nanos() as u64 / self.delay_between_requests.as_nanos() as u64;
+        self.num_requests(num_requests)
+    }
+
+    pub fn num_requests(self, num_requests: u64) -> SimulationBuilder3 {
+        let mut rng = crate::rng();
+
+        let stream = self.runtime.enter({
+            let endpoints = self.endpoints;
+            let delay_between_requests = self.delay_between_requests;
+            move || {
+                time::interval(delay_between_requests)
+                    .take(num_requests as usize)
+                    .map(move |_| endpoints.choose(&mut rng).unwrap().clone())
+            }
+        });
+
+        SimulationBuilder3 {
+            runtime: self.runtime,
+            metrics: self.metrics,
+            recorder: self.recorder,
+            builder: self.builder,
+            requests: Box::pin(stream),
+            abort_after: None,
+        }
+    }
+}
+
+pub struct SimulationBuilder3 {
+    runtime: Runtime,
+    metrics: Arc<MetricRegistry>,
+    recorder: Arc<Mutex<SimulationMetricsRecorder>>,
+    builder: Builder<SimulationRawClientBuilder>,
+    requests: Pin<Box<dyn Stream<Item = Endpoint>>>,
+    abort_after: Option<Duration>,
+}
+
+impl SimulationBuilder3 {
+    pub fn abort_after(mut self, cutoff: Duration) -> Self {
+        self.abort_after = Some(cutoff);
+        self
+    }
+
+    pub fn clients(self, clients: u32) -> Simulation {
+        let clients = (0..clients)
+            .map(|_| self.runtime.enter(|| self.builder.build().unwrap()))
+            .collect::<Vec<_>>();
+        let mut rng = crate::rng();
+
+        let client_provider = move || clients.choose(&mut rng).unwrap().clone();
+
+        Simulation {
+            runtime: self.runtime,
+            metrics: self.metrics,
+            recorder: self.recorder,
+            requests: self.requests,
+            client_provider: Box::new(client_provider),
+            abort_after: self.abort_after,
+        }
+    }
+}
+
+pub struct Simulation {
+    runtime: Runtime,
+    metrics: Arc<MetricRegistry>,
+    recorder: Arc<Mutex<SimulationMetricsRecorder>>,
+    client_provider: Box<dyn FnMut() -> Client<Arc<SimulationRawClient>>>,
+    requests: Pin<Box<dyn Stream<Item = Endpoint>>>,
+    abort_after: Option<Duration>,
+}
+
+impl Simulation {
+    pub fn builder() -> SimulationBuilder0 {
+        SimulationBuilder0
+    }
+
+    pub fn run(mut self) -> SimulationReport {
```

Comment:
nit: this method is long and deeply nested. Can we factor out some of the logic into separate functions?

Reply:
Split things up and got it flattened down nicely.